### PR TITLE
NAS-100106 / 11.3 / Take samba4 db backup before upgrade

### PR DESCRIPTION
**Pull request by themylogin**

### CODE REVIEW

File: src/middlewared/middlewared/plugins/update.py [hunk 1]
```diff
@@ -450,6 +451,15 @@ def check_available(self, attrs=None):
...
         new_manifest.LoadPath('{}/MANIFEST'.format(location))
 
+        if (
+            await self.middleware.call_sync('system.is_freenas') or
+            (
+                await self.middleware.call('failover.licensed') and
+                await self.middleware.call('failover.status') != 'BACKUP'
+            )
+        ):
+            await self.middleware.call('update.take_systemdataset_samba4_snapshot')
```
[the remark this answers is not in the document]
Comment (themylogin) [reply src/middlewared/middlewared/plugins/update.py]: Are we sure that we will never do anything with this directory contents in post-update scripts?

[the remark this answers is not in the document]
Comment (william-gr) [reply src/middlewared/middlewared/plugins/update.py]: Post-upgrade scripts run in the new BE, not over the mounted system dataset in the current BE

File: src/middlewared/middlewared/plugins/update.py [hunk 2]
```diff
@@ -645,3 +656,28 @@ def destroy_upload_location(self):
...
+            self.logger.warning('Unable to list dataset %s snapshots: %s', dataset, proc.stderr)
+            return
+
+        snapshots = [s.split('@')[1] for s in proc.stdout.strip().split()]
+        for snapshot in [s for s in snapshots if s.startswith('update-')][:-4]:
+            self.logger.info('Deleting dataset %s snapshot %s', dataset, snapshot)
+            subprocess.run(['zfs', 'destroy', f'{dataset}@{snapshot}'])
+
+        current_version = "-".join(self.middleware.call_sync("system.info")["version"].split("-")[1:])
+        new_version = "-".join(new_version.split("-")[1:])
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/update.py]: I am not seeing the usefulness of using new version here, perhaps just stick with the version its running?

Comment (yocalebo) [reply src/middlewared/middlewared/plugins/update.py]: > I am not seeing the usefulness of using new version here, perhaps just stick with the version its running?

Yeah, let's just use the current version it's running at the time of snapshot.

**Pull request merged**